Refinement of Providers into Providers and ExternProviders (#2469)
rust-lang/rustc-dev-guide: src/query.md

If, however, the query is *not* in the cache, then the compiler will
call the corresponding **provider** function. A provider is a function

> implemented in a specific module and **manually registered** into either

Reviewer: ```suggestion
pointer **manually registered** into either
```

idk what "implemented in a specific module" is about, everything is defined in a module in rust haha

### How providers are set up

When the tcx is created, it is given both the local and external providers by its creator using
the `Providers` struct from `rustc_middle::util`. This struct contains both the local and external providers:

```rust,ignore
pub struct Providers {
    pub queries: crate::query::Providers,        // Local crate providers
    pub extern_queries: crate::query::ExternProviders,  // External crate providers
    pub hooks: crate::hooks::Providers,
}
```

Each of these provider structs is generated by the macros and contains function pointers for their respective queries.

#### How are providers registered?

> The provider structs are filled in during compiler initialization, mainly by the `rustc_driver` crate.  

Reviewer: ```suggestion
The provider structs are filled in during compiler initialization, across many different `rustc_` crates.  
```
I think rustc_drive is by far not the main place that we actually initialise the provider fields to meaningful values

But the actual provider functions are implemented in various `rustc_*` crates (like `rustc_middle`, `rustc_hir_analysis`, etc).

To register providers, each crate exposes a [`provide`][provide_fn] function that looks like this:

[provide_fn]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/fn.provide.html

```rust,ignore
pub fn provide(providers: &mut rustc_middle::util::Providers) {
    providers.queries.type_of = type_of;
    // ... add more local providers here
    
    providers.extern_queries.type_of = extern_type_of;
    // ... add more external providers here
    
    providers.hooks.some_hook = some_hook;
    // ... add more hooks here

> }

Reviewer: ````suggestion
pub fn provide(providers: &mut rustc_middle::util::Providers) {
    providers.queries.type_of = type_of;
    // ... add more local providers here
    
    providers.extern_queries.type_of = extern_type_of;
    // ... add more external providers here
    
    providers.hooks.some_hook = some_hook;
    // ... add more hooks here
}
```

If only the `queries` field is being updated then it's common practice to accept `query::Providers` instead:
```rust,ignore
pub fn provide(providers: &mut Providers) {
    *providers = Providers {
        type_of,
        // ... add more providers here
        ..*providers
    };
}
```

Note that `util::Providers` implements `DerefMut` to `query::Providers` so callers of the `provide` functions can pass in a `util::Providers` and it will just work for provider functions that accept `query::Providers` too.
````

```

- This function takes a mutable reference to the `Providers` struct and sets the fields to point to the correct provider functions.

> - You can assign fields individually for each provider type (local, external, and hooks).

Reviewer: ```suggestion
- You can also assign queries individually, e.g. `providers.type_of = type_of;`.
- You can assign fields individually for each provider type (local, external, and hooks).
```

#### Adding a new provider

Suppose you want to add a new query called `fubar`. You would:

> 1. Implement the provider function:

Reviewer: ```suggestion
1. Implement the function to be provided for the query:
```

    ```rust,ignore
    fn fubar<'tcx>(tcx: TyCtxt<'tcx>, key: LocalDefId) -> Fubar<'tcx> { ... }
    ```

> 2. Register it in the `provide` function:

Reviewer: I think realistically this ought to talk about:
- try find an existing provider function in the crater
- if one doesn't exist add one (and then explain where in `rustc_interface` to actually call your new provider function from

though this is a pre-existing issue :woman_shrugging:

   - Returns the decoded information to the query system

This approach avoids recompiling external crates, allows for faster compilation of dependent crates, and enables incremental compilation to work across crate boundaries.
Here is a simplified example, when you call `tcx.type_of(def_id)` for a type defined in an external crate, the query system will:
1. Detect that the `def_id` refers to an external crate by checking `def_id.krate != LOCAL_CRATE`
2. Call the appropriate provider from `ExternProviders` which was registered by `rustc_metadata`
3. The provider will load and decode the type information from the external crate's metadata
4. Return the decoded type to the caller

This is why most `rustc_*` crates only need to provide local providers - the external providers are handled by the metadata system. 
The only exception is when a crate needs to provide special handling for external queries, in which case it would implement both local and external providers.

[rustc_metadata]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/index.html
[providers_struct]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/query/struct.Providers.html

> [extern_providers_struct]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/query/struct.ExternProviders.html

Author: Here is the new version

Here is a simplified example, when you call `tcx.type_of(def_id)` for a type defined in an external crate, the query system will:
1. Detect that the `def_id` refers to an external crate by checking `def_id.krate != LOCAL_CRATE`
2. Call the appropriate provider from `ExternProviders` which was registered by `rustc_metadata`
3. The provider will load and decode the type information from the external crate's metadata
4. Return the decoded type to the caller

This is why most `rustc_*` crates only need to provide local providers - the external providers are handled by the metadata system. 
The only exception is when a crate needs to provide special handling for external queries, in which case it would implement both local and external providers.

[rustc_metadata]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/index.html
[providers_struct]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/query/struct.Providers.html
[extern_providers_struct]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/query/struct.ExternProviders.html

> 

Author: I read the code in `rustc_metadata` related to the extern providers and added it to the new commit, but I'm not sure there are parts I wrote correctly, if you have time to look at it. Thanks @Noratrieb

#### How are providers registered?

The provider structs are filled in during compiler initialization, mainly by the `rustc_driver` crate.  
But the actual provider functions are implemented in various `rustc_*` crates (like `rustc_middle`, `rustc_hir_analysis`, etc).

To register providers, each crate exposes a [`provide`][provide_fn] function that looks like this:

[provide_fn]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/fn.provide.html

```rust,ignore
pub fn provide(providers: &mut rustc_middle::util::Providers) {
    providers.queries.type_of = type_of;
    // ... add more local providers here
    
    providers.extern_queries.type_of = extern_type_of;
    // ... add more external providers here
    
    providers.hooks.some_hook = some_hook;
    // ... add more hooks here
}
```

- This function takes a mutable reference to the `Providers` struct and sets the fields to point to the correct provider functions.
- You can assign fields individually for each provider type (local, external, and hooks).

#### Adding a new provider

Suppose you want to add a new query called `fubar`. You would:

1. Implement the provider function:
    ```rust,ignore

>     fn fubar<'tcx>(tcx: TyCtxt<'tcx>, key: LocalDefId) -> Fubar<'tcx> { ... }

Author: @tshepang suggest to make it `LocalDefId`. https://github.com/rust-lang/rustc-dev-guide/pull/2465#discussion_r2147180199

I just went through the source code and realized that the key for local providers is pretty much `LocalDefId`. Which one are we going to choose?

>     fn fubar<'tcx>(tcx: TyCtxt<'tcx>, key: LocalDefId) -> Fubar<'tcx> { ... }

Reviewer 2: only suggested that because I did not realize there is ExternProvider as well, which accepts DefId (and not LocalDefId)

either works for me... can always be adjusted later to mention both
